src/quantumchannel.jl

```diff
@@ -0,0 +1,89 @@
+"""
+QuantumChannel for transmitting qubits as `RegRef`s from one register to another in a quantum protocol simulation, 
+with a channel delay and under the influence of a background process.
+The function `put!` is used to put the `RegRef` containing the qubit in the channel, which can then be received by
+the receiving register after a specified delay using the take! method in a synchronous way.
+
+```jldoctest
+julia> using QuantumSavory; using ResumableFunctions; using ConcurrentSim
+
+julia> bell = (Z1⊗Z1 + Z2⊗Z2)/sqrt(2.0); regA = Register(2); regB = Register(2); initialize!((regA[1], regB[2]), bell);
+
+julia> sim = Simulation();
+
+julia> queue = DelayQueue{RegRef}(sim, 10.0)
+DelayQueue{RegRef}(QueueStore{RegRef, Int64}, 10.0)
+
+julia> qc = QuantumChannel(queue)
+QuantumChannel(Qubit(), DelayQueue{RegRef}(QueueStore{RegRef, Int64}, 10.0), nothing)
+
+julia> @resumable function alice_node(env, qc)
+            println("Putting Alice's qubit in the channel at ", now(env))
+            put!(qc, regA[1])
+        end
+alice_node (generic function with 1 method)
+
+julia> @resumable function bob_node(env, qc)
+            @yield @process take!(env, qc, regB[1]) # wait for the process with delay to complete
+            println("Taking the qubit from alice at ", now(env))
+        end
+bob_node (generic function with 1 method)
+
+julia> @process alice_node(sim, qc); @process bob_node(sim, qc);
+
+julia> run(sim)
+Putting Alice's qubit in the channel at 0.0
+Taking the qubit from alice at 10.0
+
+julia> regA
+Register with 2 slots: [ Qubit | Qubit ]
+  Slots:
+    nothing
+    nothing
+
+julia> regB
+Register with 2 slots: [ Qubit | Qubit ]
+  Slots:
+    Subsystem 1 of QuantumOpticsBase.Ket 12382959472027850978
+    Subsystem 2 of QuantumOpticsBase.Ket 12382959472027850978
+```
+"""
+struct QuantumChannel
+    trait::Qubit
+    queue::ConcurrentSim.DelayQueue{RegRef}
+    background::Any
+end
+
+function QuantumChannel(queue::ConcurrentSim.DelayQueue{RegRef}, background=nothing)
+    QuantumChannel(Qubit(), queue, background)
+end
+
+function Base.put!(qc::QuantumChannel, rref::RegRef, Δt=nothing)
+    if xor(isnothing(qc.background), isnothing(Δt))
+        throw(ArgumentError(lazy"""
+        Either both background and Δt should be nothing or both should be initialized to appropriate values
+        """))
+    elseif !isnothing(Δt) # if both are not nothing
+        uptotime!(rref.reg.staterefs[rref.idx], rref.reg.stateindices[rref.idx], qc.background, Δt)
+    end
+
+    put!(qc.queue, rref)
+end
+
+# should we mandate that rref_rec.reg is not initialized beforehand
+# should other register attributes of the sent regref be copied over like reprs[idx], backgrounds[idx], accesstimes[idx], env and locks[idx]
+@resumable function Base.take!(env, qc::QuantumChannel, rref_rec::RegRef)
```

Comment:
Should we check whether the register slot receiving the qubit is always uninitialized? Also, do we need to copy other information related to the qubit like the background noise associated with it, i.e., reprs(idx), backgrounds(idx), locks(idx) etc. ?

Reply:
Yes, we should check that the receiving register is empty (same check as the one currently done in `initialize!`).

And no, all of these properties should not be copied. They are properties of the location (channel or register), not properties of the state stored in that location.

Reply:
Yeah, that makes sense!